src/core/thread/mle.cpp

```diff
@@ -1419,31 +1414,28 @@ void Mle::HandleParentRequestTimer(void)
     case kParentRequestStart:
         mParentRequestState = kParentRequestRouter;
         mParentCandidate.SetState(Neighbor::kStateInvalid);
+        mReceivedResponseFromParent = false;
         SendParentRequest();
         break;
 
     case kParentRequestRouter:
         mParentRequestState = kParentRequestChild;
 
-        if (mParentCandidate.GetState() == Neighbor::kStateValid)
-        {
-            SendChildIdRequest();
-            mParentRequestState = kChildIdRequest;
-            mParentRequestTimer.Start(kParentRequestChildTimeout);
-        }
-        else
+        if (mParentCandidate.GetState() != Neighbor::kStateParentResponse)
         {
             SendParentRequest();
+            break;
         }
 
-        break;
+    // fall through
 
     case kParentRequestChild:
         mParentRequestState = kParentRequestChild;
 
-        if (mParentCandidate.GetState() == Neighbor::kStateValid)
+        if (mParentCandidate.GetState() == Neighbor::kStateParentResponse &&
+            (mRole != OT_DEVICE_ROLE_CHILD || mReceivedResponseFromParent) &&
+            SendChildIdRequest() == OT_ERROR_NONE)
```

Comment:
@jwhui  we see some incorrect behavior with the logic for the case where there is only one parent and a parent search is triggered while we are attached. We see the device get detached. In such a case, this `if` should fail (since `SendChildIdRequest()` returns `OT_ERROR_ALREADY`. 

So we go through the fall through logic below. Is this expected? Ok?

Reply:
We can also get to "fall through" logic below if we don't hear from our current parent (`mReceivedResponseFromParent` being `false`)....

Is this ok? I think we may need to add a check to see if we are already a child and break from the case below, before checking the `mReattachSatate`? Does it make sense?

Reply:
We want to fall through to handle other attach scenarios (i.e. `kAttachSame` and `kAttachBetter`), but we do want to be careful about detaching if we are already attached as a child.  See update below.